crates/matrix-sdk-crypto/src/session_manager/group_sessions/share_strategy.rs

```diff
@@ -179,73 +182,91 @@ pub(crate) async fn collect_session_recipients(
+            // out with an error.
+            if !unsigned_devices_of_verified_users.is_empty() {
+                return Err(OlmError::SessionRecipientCollectionError(
+                    SessionRecipientCollectionError::VerifiedUserHasUnsignedDevice(
+                        unsigned_devices_of_verified_users,
+                    ),
+                ));
             }
-        };
 
-        // If we're using a `DeviceBasedStrategy` with
-        // `error_on_verified_user_problem` set, then
-        // `unsigned_of_verified_user` may be populated. If so, add an entry to the
-        // list of users with unsigned devices.
-        if !recipient_devices.unsigned_of_verified_user.is_empty() {
-            unsigned_devices_of_verified_users.insert(
-                user_id.to_owned(),
-                recipient_devices
-                    .unsigned_of_verified_user
-                    .into_iter()
-                    .map(|d| d.device_id().to_owned())
-                    .collect(),
-            );
+            // Alternatively, we may have encountered previously-verified users who have
+            // changed their identities. We bail out for that, too.
+            if !verified_users_with_new_identities.is_empty() {
+                return Err(OlmError::SessionRecipientCollectionError(
+                    SessionRecipientCollectionError::VerifiedUserChangedIdentity(
+                        verified_users_with_new_identities,
+                    ),
+                ));
+            }
         }
+        CollectStrategy::IdentityBasedStrategy => {
+            for user_id in users {
+                trace!("Considering recipient devices for user {}", user_id);
+                let user_devices = store.get_device_data_for_user_filtered(user_id).await?;
 
-        let recipients = recipient_devices.allowed_devices;
-        let withheld_recipients = recipient_devices.denied_devices_with_code;
+                let device_owner_identity = store.get_user_identity(user_id).await?;
 
-        // If we haven't already concluded that the session should be
-        // rotated for other reasons, we also need to check whether any
-        // of the devices in the session got deleted or blacklisted in the
-        // meantime. If so, we should also rotate the session.
-        if !should_rotate {
-            should_rotate = is_session_overshared_for_user(outbound, user_id, &recipients)
-        }
+                let recipient_devices = split_recipients_withhelds_for_user_based_on_identity(
+                    user_devices,
+                    &device_owner_identity,
+                );
 
-        devices.entry(user_id.to_owned()).or_default().extend(recipients);
-        withheld_devices.extend(withheld_recipients);
-    }
+                let recipients = recipient_devices.allowed_devices;
+                let withheld_recipients = recipient_devices.denied_devices_with_code;
```

Comment:
consider inlining `withheld_recipients`

Reply:
Indeed, this was done in https://github.com/matrix-org/matrix-rust-sdk/pull/3662 (along with inlining `recipients`) so I've added that here now.